Extend standalone support

jerbob92 opened the pull request:
- Implement _emscripten_throw_longjmp that aborts the program
- Add nice messages to __cxa_throw and __cxa_allocate_exception
- Implement getentropy using WASI
- Implement WASI FS for syscalls lstat64, stat64, newfstatat, getdents64, fstat64 and openat

This is far from complete if you look at the WASI FS spec and the syscalls, but at least it gets basic operations like directory listing, file/directory opening and file/directory stat working. It was enough for me to get pdfium working in standalone mode.

I didn't really know what I should do with AT_FDCWD or what the runtime should do with it when passing AT_FDCWD. AT_FDCWD is also negative while I think `fd` in WASI is unsigned?

Most of the syscall code was taken from or based on wasi-libc.

sbc100 commented on system/lib/standalone/standalone.c:
```diff
@@ -155,24 +480,32 @@ double emscripten_get_now(void) {
 
 __attribute__((__weak__))
 void* __cxa_allocate_exception(size_t thrown_size) {
+  _emscripten_err("The program tried to allocate an exception, this is not supported in standalone mode, the program will exit now...");
```
Can you use macro like in `emscripten_syscall_stubs.c`: https://github.com/emscripten-core/emscripten/blob/a197a6bd0dd31cdeea60dd421dc86c1cf9a0bf80/system/lib/libc/emscripten_syscall_stubs.c#L40-L44


Adding and using that macro could be its own small change?

sbc100 replied:
These kind of messages are normally not included (by emscripten) in release builds.    I don't see why these cases should be any different.

As well as enforcing consistency the macro name also gives a some semantic issue to the reader.  Its makes it easy to search for too.

jerbob92 replied:
Alright. In my opinion you would want to make it clear why the program aborted if it's because of something that's not supported in standalone. Only doing abort doesn't really give any other info.

But I'll change it to a macro.

sbc100 replied:
Agreed that the abort on its own is now very useful, but there are lots of places in emscripten where the release build is unhelpful or less helpful, but the debug build is more helpful.    This keeps the release builds small, and (I hope) that developers are use to building a debug build to diagnose issues.

jerbob92 replied:
I added the macro.

sbc100 commented on system/lib/standalone/standalone.c:
```diff
@@ -109,15 +202,247 @@ __attribute__((__weak__)) int __syscall_fcntl64(int fd, int cmd, ...) {
   return -ENOSYS;
 }
 
+void to_public_stat(const __wasi_filestat_t *in,
```
make this `static`?

sbc100 replied:
Its not clear to me what `public` means here.   How about `from_wasi_stat`?  Or `covert_wasi_stat`?

sbc100 commented on system/lib/standalone/standalone.c:
```diff
@@ -109,15 +202,247 @@ __attribute__((__weak__)) int __syscall_fcntl64(int fd, int cmd, ...) {
-  return -ENOSYS;
+  __wasi_filestat_t internal_stat;
+  __wasi_errno_t error =
+    __wasi_fd_filestat_get(fd, &internal_stat);
```
Does this fit on a single line?

jerbob92 replied:
Fixed

sbc100 commented on system/lib/standalone/standalone.c:
```diff
@@ -109,15 +202,247 @@ __attribute__((__weak__)) int __syscall_fcntl64(int fd, int cmd, ...) {
+  }
+}
+
 __attribute__((__weak__)) int __syscall_fstat64(int fd, intptr_t buf) {
```
Elsewhere in this file it looks like we put the weak attribute on its own line.. maybe do that here to?

jerbob92 replied:
It's not really consistent over the file, but I have changed it now where the method is implemented.

sbc100 commented on system/lib/standalone/standalone.c:
```diff
@@ -195,33 +528,6 @@ void _wasmfs_get_preloaded_child_path(int index, void* buffer) {}
-void _emscripten_out(char* text) { wasi_writeln(1, text); }
-
-void _emscripten_err(char* text) { wasi_writeln(2, text); }
-
```
Why move this code up to the top?   Can we make the diff smaller by leaving it where it is?

jerbob92 replied:
Because it's being called in methods above this (`__cxa_allocate_exception`/`__cxa_throw`/`_emscripten_throw_longjmp`), keeping it at the bottom will result in errors like `call to undeclared function '_emscripten_err'`.

sbc100 replied:
But `_emscripten_out` and `_emscripten_err` are declare in the `emscripten/console.h` header so that order here shouldn't matter.

jerbob92 replied:
I had to move it because it didn't compile without it. Just tried to put it back and indeed doesn't work.
When I tried to include `emscripten/console.h` I got duplicate method definition errors. Could be my lack of C knowledge.

sbc100 replied:
I created a separate PR to address that issue: #18323.

jerbob92 replied:
Nice, thanks!

jerbob92 replied:
I merged in https://github.com/emscripten-core/emscripten/pull/18323 to fix the macro, it also needs the console error to be defined.

sbc100 reviewed (commented):
Thanks for working on this!

I left a few nits.  Perhaps we can maybe split this up.

sbc100 commented:
Can we add some some new tests based on these?

jerbob92 commented:
> Can we add some some new tests based on these?

I have never developed in C before, I don't really know where to start, can you give me some hints where you want these tests added and what they should test?

sbc100 commented on system/lib/standalone/standalone.c:
```diff
@@ -32,6 +38,13 @@
+#define REPORT_UNSUPPORTED(action)
+#else
+#define REPORT_UNSUPPORTED(action) \
+  emscripten_console_error("the program tried to " #action ", this is not supported in standalone mode, the program will exit now...\n");
```
You don't need the tailing newline here.

Also, maybe just drop the `, the program will exit now...` part.

jerbob92 replied:
Updated

jerbob92 commented on system/lib/standalone/standalone.c:
```diff
@@ -155,24 +462,32 @@ double emscripten_get_now(void) {
 
+__attribute__((__weak__))
+void _emscripten_throw_longjmp() {
+  REPORT_UNSUPPORTED(do an invalid longjmp);
```
I think I could only trigger this by doing an invalid jump, but I think you would know better.

jerbob92 replied:
Thanks! Not that it matters but do you have any C to demonstrate this? Just curious.
In my tests I was only able to trigger this by doing an invalid jmp. I think the jumps were being handled by the invoke_* methods?

sbc100 replied:
Any call to `longjmp` from C code will called `_emscripten_throw_longjmp` IIUC.

So here I would just do `REPORT_UNSUPPORTED(longjmp)`

sbc100 commented on system/include/wasi/api.h:
```diff
@@ -1008,6 +1008,19 @@ typedef uint32_t __wasi_lookupflags_t;
+ *   - https://pubs.opengroup.org/onlinepubs/9699919799/functions/V2_chap02.html#tag_15_14
+ *   - https://github.com/WebAssembly/wasi-libc/blob/wasi-sdk-16/libc-bottom-half/sources/preopens.c#L215
+ */
+#define __WASI_FD_ROOT 3
```
I believe the way it works is that there can be any number of pre-opens, starting at 3: https://github.com/WebAssembly/wasi-libc/blob/30094b6ed05f19cee102115215863d185f2db4f0/libc-bottom-half/sources/preopens.c#L212-L215.   And non of these pre-opens are necessarily the root, they can be mounted in various places.

Also, can you move this macros (if you keep it), into standalone.c.   api.h is copied from wasi-libc, and I think its better to avoid have local modifications to it.

sbc100 reviewed (commented):
I think we really needs some tests for this.   

The most basic test for the standalone syscalls is in `test_standalone_syscalls` in `test_other.py`.

For other tests look for `@also_with_standalone_wasm` in `test_core.py`

codefromthecrypt commented:
on behalf of wasm people I know I would like to thank @jerbob92 and @sbc100 for progressing this, as it not only unlocks things like pdfium, but also generic wasm utilities that use emscripten (like wabt). Ex being able to run wat2wasm using any wasi runtime instead of whatever was released and packaged.

If any one not currently involved has skills to contribute, please do as this is very likely well into "hobby time" for @jerbob92 and it is at the core a pretty substantial infrastructure change and a lot of work to test it. other wasm people will thank you, but I'll thank in advance!

sbc100 commented:
> on behalf of wasm people I know I would like to thank @jerbob92 and @sbc100 for progressing this, as it not only unlocks things like pdfium, but also generic wasm utilities that use emscripten (like wabt). Ex being able to run wat2wasm using any wasi runtime instead of whatever was released and packaged.

If you just wast a wasi version of the wabt tools I think bon behalf of wasm people I know I would like to thank @jerbob92 and @sbc100 for progressing this, as it not only unlocks things like pdfium, but also generic wasm utilities that use emscripten (like wabt). Ex being able to run wat2wasm using any wasi runtime instead of whatever was released and packaged.

If any one not currently involved has skills to contribute, please do as this is very likely well into "hobby time" for @jerbob92 and it is at the core a pretty substantial infrastructure change and a lot of work to test it. other wasm people will thank you, but I'll thank in advance!uilding them with wasi-sdk is probably simple.   That doesn't mean we shouldn't land this change of course.

> 
> If any one not currently involved has skills to contribute, please do as this is very likely well into "hobby time" for @jerbob92 and it is at the core a pretty substantial infrastructure change and a lot of work to test it. other wasm people will thank you, but I'll thank in advance!

codefromthecrypt commented:
@sbc100 ps your last comment didn't format right. might want to polish it

codefromthecrypt commented:
> If you just wast a wasi version of the wabt tools I think the simplest path would be to use wasi-sdk to build it. Of course that doesn't mean we shouldn't land this change too.

good idea, on the wabt thing and indeed a separate topic if they are open to decoupling from emscripten.

sbc100 commented:
(I guess maybe what you are saying is that it would be nice if the web version of wabt, which is run by emscripten, happened to also be WASI compliant.. in that case I agree that would be very cool).

codefromthecrypt commented:
yeah sorry it was that we were looking for a wabt compiled to wasm and found the one that was compiled with emscripten for the web. I raised this issue for out of browser wasi binary https://github.com/WebAssembly/wabt/issues/2101

jerbob92 commented:
Thanks @codefromthecrypt! So in hindsight, I don't think my C skills are good enough to get this PR into a merge-able state, implementing AT_FDCWD, using the preload dirs and not the FD 3, implementing tests.

codefromthecrypt commented:
@jerbob92 no worries, I think you got things very far. I'll keep recruiting to whatever end on this.

jerbob92 commented:
@sbc100 I have done some more progress on this PR now:
 - Implemented WASI pre-opens
 - Implement mkdirat

I do have some questions though:
 - How to get started on adding tests for this?
 - The standalone.c file is getting very big right now, how do we want to split this up? Perhaps something like standalone_wasi_preopens.c, standalone_wasi_fs.c, standalone_wasi_random.c go keep things clean?
 - Is there an easy way to check which syscalls still need to be implemented in standalone to make it full WASI compliant?
 - Code is mostly copied from wasi-libc right now with some changes, is this something that we want?
 - wasi-libc has a whole system to maintain a CWD, is this also something that we want to implement?

sbc100 commented:
> @sbc100 I have done some more progress on this PR now:
> 
> * Implemented WASI pre-opens
> * Implement mkdirat
> 
> I do have some questions though:
> 
> * How to get started on adding tests for this?
> * The standalone.c file is getting very big right now, how do we want to split this up? Perhaps something like standalone_wasi_preopens.c, standalone_wasi_fs.c, standalone_wasi_random.c go keep things clean?

I don't feel strongly about this.  Perhaps use the same split that wasi-libc uses?   Also I don't think we need `standalone` in each filename since they will all be in the `standalone` directory.


> * Is there an easy way to check which syscalls still need to be implemented in standalone to make it full WASI compliant?

The ideas is that we will be running the full wasi testsuite.   I started adding some of it in #12704 and I have some
more plans.

> * Code is mostly copied from wasi-libc right now with some changes, is this something that we want?

I don't think its a problem, but please document the origin and try to document any changes from upstream.

> * wasi-libc has a whole system to maintain a CWD, is this also something that we want to implement?

Sure.  We could even consider adding wasi-libc as a submodule and including certain files directly?

jerbob92 commented:
> I don't feel strongly about this. Perhaps use the same split that wasi-libc uses? Also I don't think we need standalone in each filename since they will all be in the standalone directory.

Sounds good!

> The ideas is that we will be running the full wasi testsuite. I started adding some of it in https://github.com/emscripten-core/emscripten/pull/12704 and I have some
more plans.

Nice, that will make it a lot easier!

> I don't think its a problem, but please document the origin and try to document any changes from upstream.
Sure. We could even consider adding wasi-libc as a submodule and including certain files directly?

If that's a possibility that would be great, it would make it a lot easier because right now I'm just cherry-picking wasi-libc code to get the syscalls/WASI calls working that I require, especially if we want to pass the whole wasi-testsuite then we have to copy a lot from was-libc. Only thing I noticed while copying code is that their WASI function signature is a bit different sometimes, which prevents us from directly using the C code (AFAIK), so we might want to look into making that possible. For example:
wasi-libc `__wasi_path_filestat_get` = `__wasi_errno_t __wasi_path_filestat_get(__wasi_fd_t fd, __wasi_lookupflags_t flags, const char *path, __wasi_filestat_t *retptr0)`
Emscripten `__wasi_path_filestat_get` = `__wasi_errno_t __wasi_path_filestat_get(__wasi_fd_t fd, __wasi_lookupflags_t flags, const char *path, size_t path_len, __wasi_filestat_t *buf)`

So Emscripten requires the path length and wasi-libc doesn't. But perhaps we can work around this by completely using wasi-libc for the wasi part.

sbc100 commented on system/lib/standalone/standalone.c:
```diff
@@ -563,18 +563,110 @@ __attribute__((__weak__)) int __syscall_ioctl(int fildes, int request, ...) {
+  return 0;
+}
+
+int rmdirat(int dirfd, intptr_t path) {
```
Can you make this `static` and make it clear this is an internal helper?

karelbilek commented:
A stupid nit - if it was based on wasi libc, shouldn't there be copyright attribution somewhere? (Or is that not necessary?)

sbc100 commented:
> A stupid nit - if it was based on wasi libc, shouldn't there be copyright attribution somewhere? (Or is that not necessary?)

No stupid at all, we should certainly consider how to handle this.  I'm not sure why the right answer is.  Or how we should deal with keeping the codebases in sync.   One easy option wold be to add wasi-libc as a submodule instead of duplicating the code here.

jerbob92 committed:
Reapply changes to 3.1.37

jerbob92 committed:
Fix duplicate standalone

jerbob92 committed:
Merge in 3.1.42

jerbob92 committed:
Fix func signature

jerbob92 committed:
Update to 3.1.44

jerbob92 committed:
Remove .idea

jerbob92 committed:
Remove .idea

jerbob92 committed:
Remove .idea

jerbob92 committed:
Throw out _emscripten_throw_longjmp stub

jerbob92 committed:
Merge tag '3.1.45' into implement-more-syscalls-for-standalone

jerbob92 committed:
Merge tag '3.1.46' into implement-more-syscalls-for-standalone

jerbob92 committed:
Merge tag '3.1.47' into implement-more-syscalls-for-standalone

d3lm commented:
Is there still the idea to get this merged at some point and maintain two ABIs?

sbc100 commented:
> Is there still the idea to get this merged at some point and maintain two ABIs?

It hasn't been big priority recently since folks who want standalone support have generally been choosing to use wasi-sdk.  Can you describe your use case for wanting better standalone support in emscripten over using wasi-sdk?  It might help motivate the direction here.

karelbilek commented:
@jerbob92 will probably repond, but for me - wasi-sdk is hard to use, because emscripten already has ready versions of stuff like freetype, zlib, libpng, etc.

with @jerbob92 we were building ghostscript and pdftops to WASI, it's "easy" with emscripten (easy in big quotes) but I don't know how would I even start with wasi-sdk with all the freetype libs etc.

jerbob92 commented:
Yeah it's basically what @karelbilek said. Most existing applications/libraries I can just compile with Emscripten, which I can't with wasi-sdk, at least not with a lot of effort.

sbc100 commented:
> Yeah it's basically what @karelbilek said. Most existing applications/libraries I can just compile with Emscripten, which I can't with wasi-sdk, at least not with a lot of effort.

Oh I didn't realize that.  I would expect most stuff to just compile with wasi-sdk.

d3lm commented:
> Can you describe your use case for wanting better standalone support in emscripten over using wasi-sdk?

I think the primary use case would maybe to have some hybrid target where you can target WASI where possible and otherwise fallback to emscripten implementations to also support APIs which aren't supported by WASI. I don't know if that's entirely out of scope and if it would be an "either or" but at least that's what I had in mind.

d3lm commented:
> I would expect most stuff to just compile with wasi-sdk.

Yea that is unfortunately not the case 😭

Emscripten just includes a lot of batteries and it'd be really nice to use WASI for example for fs syscalls because it's more efficient than going through the Node.js FS layer and then for other APIs use Emscripten implementations.

sbc100 commented:
> > they couldn't match the emscripten custom FS APIs on performance and code size in all cases
> 
> Oh wow, that is very interesting. I would have expected that the WASI FS APIs were more performant than going through Node's JS layer for FS calls.

I'm not sure what you mean here by WASI FS APIs.  Are you talking about a particular implementation?     In emscripten the WASI APIs are implemented in JS just like all the other APIs we have: https://github.com/emscripten-core/emscripten/blob/main/src/library_wasi.js.  

Perhaps you are referring to some kind of native WASI API that exists on native wasm VMs?  Or are you talking some kind of node-provided implementation of WASI?    uvwasi maybe?

> 
> Do you have an idea why that is? Where does the performance come from with Emscripten's custom FS APIs? Isn't that FS API just going through Node's `fs` layer?

IIRC it was that the ergonomics of the various APIs such as `pathopen` had overheads that emscripten syscall later didn't have or could work around.   The emscripten syscall layer is modeled after the linux syscall layer used in musl.   @kripken can you remember the specific performance issue that stopped us moving foward with switch to WASI internally by default?

karelbilek commented:
Just to explain a bit what we are doing (well mostly what @jerbob92  is doing :D ) - we build C code base with emscripten with this patch, then pass that to wazero (go WASI implementation). It mostly works.

https://github.com/tetratelabs/wazero

sbc100 commented:
> Just to explain a bit what we are doing (well mostly what @jerbob92 is doing :D ) - we build C code base with emscripten with this patch, then pass that to wazero (go WASI implementation). It mostly works.
> 
> https://github.com/tetratelabs/wazero


That makes sense.  Do you know if wazero also implements some of the emscripten syscalls?  Or is it only WASI syscalls?

kripken commented:
@sbc100 

See https://v8.dev/blog/emscripten-standalone-wasm#necessary-api-differences regarding the overhead and other issues that prevented us from moving emscripten to use 100% WASI APIs. But perhaps WASI has improved since then?

jerbob92 commented:
> > Just to explain a bit what we are doing (well mostly what @jerbob92 is doing :D ) - we build C code base with emscripten with this patch, then pass that to wazero (go WASI implementation). It mostly works.
> > https://github.com/tetratelabs/wazero
> 
> That makes sense. Do you know if wazero also implements some of the emscripten syscalls? Or is it only WASI syscalls?

Wazero implements [WASI](https://github.com/tetratelabs/wazero/tree/main/internal/wasip1) and some [Emscripten specific things](https://github.com/tetratelabs/wazero/blob/main/internal/emscripten/emscripten.go) (emscripten_notify_memory_growth, _emscripten_throw_longjmp and invoke_xxx functions). I have also implemented C++ exceptions support, but I'm still [waiting for some feedback for that](https://github.com/tetratelabs/wazero/issues/1654).

Embind support is possible using [wazero-emscripten-embind](https://github.com/jerbob92/wazero-emscripten-embind).

Since no other host calls have been implemented, most normal Emscripten builds won't work on Wazero, which is why I added more support for standalone builds. I think it makes sense that Wazero focusses on WASI and not on implementing specific Emscripten host calls, also because they are prone to change between Emscripten versions.

d3lm commented:
@sbc100 Sorry for being vague. Yea I was referring to Node's WASI implementation for instance via `uvwasi` and the implementations for the WASI interface is coming from the internal bindings so they are natively implemented in C++.

> but it won't be in the name of more efficiency

That I still don't fully understand. Why would it be less efficient to use native APIs at least when running on Node.js?

sbc100 commented:
> @sbc100 Sorry for being vague. Yea I was referring to Node's WASI implementation for instance via `uvwasi` and the implementations for the WASI interface is coming from the internal bindings so they are natively implemented in C++.

I see, yes that could be conceivable more efficient.    However with emscripten-generated JS we implement WASI in JS since we have our own virtual file system.  i.e. the JS normally has run on the web as well as under node so it can't /doesn't assume uvwasi support.

Perhaps you are right that if we push forward with STANDALONE mode and PURE_WASI mode then we could potentially bind uvwasi directly to the wasm module.  Especially for users that specify NODERAWFS (i.e. user that already opt out of emscripten's virtual filesystem).

However,  given that we have limited time to work on stuff, and given that the direction PURE_WASI seems to be heading in is to essentially copy or import wasi-libc into emscripten I'm inclined to suggest that suggesting folks simply use wasi-sdk where possible.

sbc100 commented:
> > @sbc100 Sorry for being vague. Yea I was referring to Node's WASI implementation for instance via `uvwasi` and the implementations for the WASI interface is coming from the internal bindings so they are natively implemented in C++.
> 
> I see, yes that could be conceivable more efficient. However with emscripten-generated JS we implement WASI in JS since we have our own virtual file system. i.e. the JS normally has run on the web as well as under node so it can't /doesn't assume uvwasi support.
> 
> Perhaps you are right that if we push forward with STANDALONE mode and PURE_WASI mode then we could potentially bind uvwasi directly to the wasm module. Especially for users that specify NODERAWFS (i.e. user that already opt out of emscripten's virtual filesystem).
> 
> However, given that we have limited time to work on stuff, and given that the direction PURE_WASI seems to be heading in is to essentially copy or import wasi-libc into emscripten I'm inclined to suggest that suggesting folks simply use wasi-sdk where possible.

Having said that I will try to take another look at this PR today and see if we can get it landed !

jerbob92 committed:
Merge tag '3.1.48' into implement-more-syscalls-for-standalone

d3lm commented:
> However with emscripten-generated JS we implement WASI in JS since we have our own virtual file system. i.e. the JS normally has run on the web as well as under node so it can't /doesn't assume uvwasi support.

I see. That's a fair point.

> Perhaps you are right that if we push forward with STANDALONE mode and PURE_WASI mode then we could potentially bind uvwasi directly to the wasm module. Especially for users that specify NODERAWFS (i.e. user that already opt out of emscripten's virtual filesystem).

Yea that's what I was thinking if users opt-in to use Node's FS instead of Emscripten's virtual fs.

> Having said that I will try to take another look at this PR today and see if we can get it landed!

🎉 🙏

jerbob92 committed:
Merge tag '3.1.50' into implement-more-syscalls-for-standalone

agnivade commented:
Just adding another vote of support to this PR. Currently, I am in the process of compiling [whisper.cpp](https://github.com/ggerganov/whisper.cpp) to wasi, and without this PR, it's impossible for me to access the model files from the filesystem. Would highly appreciate if this got merged into main.

sbc100 commented:
> Just adding another vote of support to this PR. Currently, I am in the process of compiling [whisper.cpp](https://github.com/ggerganov/whisper.cpp) to wasi, and without this PR, it's impossible for me to access the model files from the filesystem. Would highly appreciate if this got merged into main.

Can you explain more about your use case?  Is there some reason wasi-sdk doesn't work for you?  What platform/engines are you targetting?

agnivade commented:
> Off topic but are you sure you want to be bypassing libc like that? Wouldn't that mean re-implementing a lot of the stuff the libc provides? What do you gain from that?

Yeah, this is going slightly off-topic and into the internals of Go, but yes Go has the ability to generate builds that do not depend on libc at all. It has alternate implementations of all libc functionality. The gain is portability. Our software is used in a large variety of customer environments on a plethora of linux flavors (sometimes on very old libc versions). Go gives us the ability to not depend on libc and I'd like to use that leverage.

> Where are you hoping to run the you code? Are you targeting JS environments such as the browser or node?

Server side. Again, all of this is purely in Go. Go generates single binaries that you can simply run. 

>  I'm a little confused since pretty much all of the advanced features of emscripten (such as bindings) rely heavily on JavaScript and non-WASI APIs.

Yes, as mentioned earlier in this thread, there are libraries which re-implement all the bindings in pure Go, allowing users to use emscripten generated builds in a pure Go environment. Refer to: https://github.com/emscripten-core/emscripten/pull/18285#issuecomment-1815890406

Hope that was clear. Happy to answer any further questions :) I fully realize that emscripten is targeted towards JS based environments. But there's a lot of tooling already built that gets us to 90% of what we want to get it to run on non-JS environments. This small PR would almost take it to 99%. Ultimately, it's your decision of course as a maintainer. But this would simply unlock a lot of functionality that would otherwise require a lot of other workarounds.

agnivade commented:
> Sounds like your libc re-implementations would then be built on top of the syscalls layer which is then, by definition, not portable between OSes.. meaning you would need to re-implement for each OS. But I guess you decided to trade off those two forms of non-portableness?

Correct. Go has proper cross-compilation support and automatically takes care of this. 

> For example, if you are already implementing the emscripten ABI/API, why not just continue to do that.. 

Oh yeah we could. But from what I understand, the `openat` and other similar functions need to be exposed for us to re-implement it. @jerbob92 can probably provide more details as to why it's possible to do this from Go. Currently, only these can be overridden:

```
  (import "wasi_snapshot_preview1" "clock_time_get" (func $__wasi_clock_time_get (type 31)))
  (import "wasi_snapshot_preview1" "proc_exit" (func $__wasi_proc_exit (type 3)))
  (import "wasi_snapshot_preview1" "fd_write" (func $__wasi_fd_write (type 10)))
  (import "wasi_snapshot_preview1" "fd_read" (func $__wasi_fd_read (type 10)))
  (import "wasi_snapshot_preview1" "fd_close" (func $__wasi_fd_close (type 0)))
  (import "wasi_snapshot_preview1" "fd_seek" (func $__wasi_fd_seek (type 66)))
  (import "wasi_snapshot_preview1" "environ_sizes_get" (func $__wasi_environ_sizes_get (type 1)))
  (import "wasi_snapshot_preview1" "environ_get" (func $__wasi_environ_get (type 1)))
```

sbc100 commented:
> IMHO it's way better to add support for WASI in Emcripten than to add Emscripten ABI/API support for every runtime out there.

Perhaps, but I think that argument only really makes sense if the resulting binary is WASI-compaible.  In this case it sounds like the resulting binary will contain a bunch of emscripten-specific stuff anyway, so won't be  WASI-compliant moulde, right?  (or am I misunderstanding?)

jerbob92 commented:
> > IMHO it's way better to add support for WASI in Emcripten than to add Emscripten ABI/API support for every runtime out there.
> 
> Perhaps, but I think that argument only really makes sense if the resulting binary is WASI-compaible. In this case it sounds like the resulting binary will contain a bunch of emscripten-specific stuff anyway, so won't be WASI-compliant moulde, right? (or am I misunderstanding?)

Correct, there are still a few things that are needed to run Emscripten binaries:

1.  Embind adds a lot of Emscripten imports, but you don't have to use Embind so I don't think that's a problem
2. `emscripten_notify_memory_growth`
3. `_emscripten_throw_longjmp`
4. `invoke_xxx` generated methods for indirect function calls and catching errors

`emscripten_notify_memory_growth` could probably be removed in standalone mode (don't know if it adds any value there, in Wazero it doesn't do anything). I suspect that 3 and 4 can be get rid of when more features of WASI get to an implementation phase.

jerbob92 committed:
Merge tag '3.1.51' into implement-more-syscalls-for-standalone

jerbob92 committed:
Merge tag '3.1.53' into implement-more-syscalls-for-standalone

jerbob92 committed:
Merge tag '3.1.56' into implement-more-syscalls-for-standalone

jerbob92 committed:
Merge in 3.1.63

jerbob92 committed:
Add missing functions in WASI

jerbob92 committed:
Merge in 3.1.72

kripken commented:
@sbc100 What do you think about moving forward with this PR? It should at least help in some cases. E.g. one of the syscalls missing in #23008 is implemented here (and perhaps more, I'm not sure by the names).

sbc100 commented:
Sure, I'd be OK with moving forward with this.  I think it still needs a bit of work though.

jerbob92 committed:
Merge tag '4.0.3' into implement-more-syscalls-for-standalone

turbolent commented:
Would it be possible to create a TODO list for the work that still need to be done to get this PR merged? I'd be happy to help, but not quite sure where to start. Thanks!

jerbob92 committed:
Merge tag '4.0.7' into implement-more-syscalls-for-standalone

jerbob92 committed:
Add time related libc

jerbob92 committed:
Update time implementations for patch

karelbilek commented:
I want to say I am using this branch in building ghostscript (both the
ancient GPL version and the current AfferoGPL version) and Xpdf comand line
tools (latest version) and it all works as expected in wazero (go WASI
runtime).


On Sat 15. 3. 2025 at 20:48, Bastian Müller ***@***.***>
wrote:

> Would it be possible to create a TODO list for the work that still need to
> be done to get this PR merged? I'd be happy to help, but not quite sure
> where to start. Thanks!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/emscripten-core/emscripten/pull/18285#issuecomment-2726964075>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAZT4NUVKP4NLMZPWFB5ST2UR7XFAVCNFSM6AAAAAASQXFBISVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOMRWHE3DIMBXGU>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
> [image: turbolent]*turbolent* left a comment
> (emscripten-core/emscripten#18285)
> <https://github.com/emscripten-core/emscripten/pull/18285#issuecomment-2726964075>
>
> Would it be possible to create a TODO list for the work that still need to
> be done to get this PR merged? I'd be happy to help, but not quite sure
> where to start. Thanks!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/emscripten-core/emscripten/pull/18285#issuecomment-2726964075>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAZT4NUVKP4NLMZPWFB5ST2UR7XFAVCNFSM6AAAAAASQXFBISVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOMRWHE3DIMBXGU>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

kripken commented:
Concretely I think what is needed here is

* Add test to cover the new functionality
* Get CI fully passing

Anything else @sbc100 ?

jiixyj commented on system/lib/standalone/standalone.c:
```diff
@@ -65,53 +81,1009 @@ weak int _munmap_js(
+    case O_RDONLY:
+    case O_RDWR:
+    case O_WRONLY:
+      if ((flags & O_RDONLY) != 0) {
```
Thanks for your work on this so far!

Musl defines `O_RDONLY` as `0`, so this line will not work I think. This will only work in implementations where the file access modes are single bits, one for each file access mode (like wasi-libc, where this is the case).

The robust, POSIX compliant way to deal with file access modes is to:

1. first mask the file access modes part off with `flags & O_ACCMODE`
2. then compare the resulting value directly to the five defined access modes (`O_RDONLY`, `O_WRONLY`, `O_RDWR`, `O_SEARCH`, `O_EXEC`)

With this, it should be possible to open a file as `O_RDONLY`.

Oh, and if you test with Wasmer: Right now, it doesn't support opening a file with `O_RDONLY` (<https://github.com/wasmerio/wasmer/issues/4892>). This took me quite a while to figure out.

jerbob92 replied:
[This code comes from wasi-libc](https://github.com/WebAssembly/wasi-libc/blob/e9524a0980b9bb6bb92e87a41ed1055bdda5bb86/libc-bottom-half/cloudlibc/src/libc/fcntl/openat.c#L23), which also uses musl, so then it would be broken there as well?

jiixyj replied:
I don't know exactly how much they use from musl, but they defined the file access mode bits themselves at least:

https://github.com/WebAssembly/wasi-libc/blob/e9524a0980b9bb6bb92e87a41ed1055bdda5bb86/libc-bottom-half/headers/public/__header_fcntl.h#L18-L22

jerbob92 replied:
Hmm, yeah they defined it both at the top and bottom half (with different values). I think we will known soon enough if this works or not if we add tests.

I'm mostly just copy pasting stuff from wasi-libc here, so I wouldn't know if it's correct, it works for me but I'm not sure if I hit this code path.

jiixyj commented on system/lib/standalone/standalone.c:
```diff
@@ -65,53 +81,1009 @@ weak int _munmap_js(
+
+    // If we can't find a preopen for it, fail as if we can't find the path.
+    if (dirfd == -1) {
+      errno = ENOENT;
```
I think this should be `return -ENOENT`. And similar elsewhere. In wasi-libc, this code is used in function that use the "return -1 + errno" error convention, while the `__syscall_*` functions use the "return negative error" convention of the Linux syscall API.

jiixyj commented:
@jerbob92 : Are you or is anyone else working on the test integration at the moment? Based on this PR I managed to get the `./test/fcntl/test_fcntl_open.c` test working and hooked up using the `@also_with_standalone_wasm` test decorator.

I used this test because it is pretty simple. It just needed `open`, `stat`, `mkdir` and `symlink`.

Should I open a new PR?

jerbob92 commented:
@jiixyj Not that I know of! New MR is fine, you can also update this one if you have access.

jiixyj commented:
Based on code from this PR and wasm-libc, I've opened PRs https://github.com/emscripten-core/emscripten/pull/24246 and https://github.com/emscripten-core/emscripten/pull/24247 if anyone wants to have a look.

I have mostly concentrated on the filesystem stuff, so not all syscalls from this PR are included. But it shouldn't be too hard to add them in follow up PRs.

jerbob92 committed:
Merge in 4.0.20

jerbob92 committed:
Remove idea

jerbob92 committed:
Merge in 5.0.5

jerbob92 committed:
Merge in 6.0.0

realoriginal commented:
What is the status of this? Anyway I could help get this merged?

kripken commented:
I think the overall status is that this would be good to land, and the next step is testing (test the new functionality, and get all the existing tests passing).

sbc100 commented on system/lib/standalone/standalone.c:
```diff
@@ -337,6 +1338,252 @@ weak void _emscripten_get_progname(char* buf, int length) {
+		errno = EOVERFLOW;
+		return 0;
+	}
+	__secs_to_zone(*t, 0, &tm->tm_isdst, &tm->__tm_gmtoff, 0, &tm->__tm_zone);
```
I'm a little confused by the call to `__secs_to_zone` since we declare this function as unsupported just above?

jerbob92 replied:
I don't recall. I only updated this to use the latest Emscripten in my compilations, this has been in there for a while, I think this is just copied from wasi-libc as well.

jerbob92 replied:
@sbc100 Maybe it's better to just close this PR?
